File: mtags-interfaces/src/main/java/scala/meta/pc/ReferencesRequest.java
package scala.meta.pc;

import java.util.Collections;
import java.util.List;
import org.eclipse.lsp4j.jsonrpc.messages.Either;

public interface ReferencesRequest {
  VirtualFileParams file();

  boolean includeDefinition();

  Either<Integer, String> offsetOrSymbol();

  default List<String> alternativeSymbols() {
    return Collections.emptyList();
  }
}

File: mtags-shared/src/main/scala/scala/meta/internal/pc/ReferencesRequest.scala
package scala.meta.internal.pc

import java.{util => ju}

import scala.collection.JavaConverters._

import scala.meta.pc.ReferencesRequest
import scala.meta.pc.ReferencesResult
import scala.meta.pc.VirtualFileParams

import org.eclipse.lsp4j.Location
import org.eclipse.lsp4j.jsonrpc.messages.{Either => JEither}

case class PcReferencesRequest(
    file: VirtualFileParams,
    includeDefinition: Boolean,
    offsetOrSymbol: JEither[Integer, String],
    override val alternativeSymbols: ju.List[String] = Nil.asJava
) extends ReferencesRequest

case class PcReferencesResult(
    symbol: String,
    locations: ju.List[Location]
) extends ReferencesResult

object PcReferencesResult {
  def empty: ReferencesResult =
    PcReferencesResult("", ju.Collections.emptyList())
}
